File: cmd/update_check.go
package cmd

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/spf13/cobra"
)

const (
	updateCheckRepo        = "VantaInc/cli"
	updateCheckInterval    = 24 * time.Hour
	updateCheckHTTPTimeout = 1500 * time.Millisecond
	updateCheckEnvDisable  = "VANTA_NO_UPDATE"
	installScriptURL       = "https://raw.githubusercontent.com/VantaInc/cli/main/scripts/install.sh"
)

type updateCheckCache struct {
	CheckedAt     time.Time `json:"checked_at"`
	LatestVersion string    `json:"latest_version"`
}

type githubLatestRelease struct {
	TagName string `json:"tag_name"`
}

var (
	updateHTTPClient = &http.Client{Timeout: updateCheckHTTPTimeout}
	updateCheckNow   = time.Now

	updateCheckState = struct {
		once    sync.Once
		mu      sync.Mutex
		started bool
		done    chan struct{}
		latest  string
		err     error
	}{
		done: make(chan struct{}),
	}
)

func shouldNotifyUpdates(cmd *cobra.Command) bool {
	if normalizeVersion(Version) == "" || Version == "dev" {
		return false
	}
	if envTruthy(os.Getenv(updateCheckEnvDisable)) {
		return false
	}
	if envTruthy(os.Getenv("CI")) {
		return false
	}
	if agentModeEnabled(cmd) {
		return false
	}
	if !isTerminal(os.Stderr) {
		return false
	}
	return true
}

func envTruthy(value string) bool {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func isTerminal(f *os.File) bool {
	if f == nil {
		return false
	}
	info, err := f.Stat()
	if err != nil {
		return false
	}
	return info.Mode()&os.ModeCharDevice != 0
}

func startBackgroundUpdateCheck(cmd *cobra.Command) {
	if !shouldNotifyUpdates(cmd) {
		return
	}

	updateCheckState.once.Do(func() {
		updateCheckState.mu.Lock()
		updateCheckState.started = true
		updateCheckState.mu.Unlock()

		go func() {
			defer close(updateCheckState.done)
			latest, err := resolveLatestVersion()
			updateCheckState.mu.Lock()
			updateCheckState.latest = latest
			updateCheckState.err = err
			updateCheckState.mu.Unlock()
		}()
	})
}

func finishBackgroundUpdateCheck(cmd *cobra.Command) {
	updateCheckState.mu.Lock()
	started := updateCheckState.started
	updateCheckState.mu.Unlock()
	if !started {
		return
	}

	select {
	case <-updateCheckState.done:
	case <-time.After(updateCheckHTTPTimeout):
		return
	}

	updateCheckState.mu.Lock()
	latest := updateCheckState.latest
	err := updateCheckState.err
	updateCheckState.mu.Unlock()
	if err != nil || latest == "" {
		return
	}
	if !isNewerVersion(latest, Version) {
		return
	}

	fmt.Fprintf(
		cmd.ErrOrStderr(),
		"\nA new version of vanta is available: %s (you have %s)\nUpdate with:\n  curl -fsSL %s | bash\n\n",
		displayVersion(latest),
		displayVersion(Version),
		installScriptURL,
	)
}

func resolveLatestVersion() (string, error) {
	if cached, ok, err := readUpdateCheckCache(); err == nil && ok {
		if updateCheckNow().Sub(cached.CheckedAt) < updateCheckInterval && cached.LatestVersion != "" {
			return cached.LatestVersion, nil
		}
	}

	latest, err := fetchLatestReleaseVersion()
	if err != nil {
		return "", err
	}

	_ = writeUpdateCheckCache(updateCheckCache{
		CheckedAt:     updateCheckNow(),
		LatestVersion: latest,
	})
	return latest, nil
}

func fetchLatestReleaseVersion() (string, error) {
	req, err := http.NewRequest(http.MethodGet, "https://api.github.com/repos/"+updateCheckRepo+"/releases/latest", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", userAgent())

	resp, err := updateHTTPClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return "", err
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("github releases latest: %s", resp.Status)
	}

	var release githubLatestRelease
	if err := json.Unmarshal(body, &release); err != nil {
		return "", err
	}
	tag := strings.TrimSpace(release.TagName)
	if tag == "" {
		return "", fmt.Errorf("github releases latest: empty tag_name")
	}
	return normalizeVersion(tag), nil
}

func updateCheckCachePath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".vanta", "update-check.json"), nil
}

func readUpdateCheckCache() (updateCheckCache, bool, error) {
	path, err := updateCheckCachePath()
	if err != nil {
		return updateCheckCache{}, false, err
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return updateCheckCache{}, false, nil
		}
		return updateCheckCache{}, false, err
	}

	var cached updateCheckCache
	if err := json.Unmarshal(raw, &cached); err != nil {
		return updateCheckCache{}, false, err
	}
	return cached, true, nil
}

func writeUpdateCheckCache(cached updateCheckCache) error {
	path, err := updateCheckCachePath()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(cached, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(path, raw, 0o600)
}

func isNewerVersion(latest, current string) bool {
	latestParts, latestOK := parseSemver(normalizeVersion(latest))
	currentParts, currentOK := parseSemver(normalizeVersion(current))
	if !latestOK || !currentOK {
		return normalizeVersion(latest) != normalizeVersion(current) && normalizeVersion(latest) > normalizeVersion(current)
	}
	for i := 0; i < 3; i++ {
		if latestParts[i] > currentParts[i] {
			return true
		}
		if latestParts[i] < currentParts[i] {
			return false
		}
	}
	return false
}

func parseSemver(version string) ([3]int, bool) {
	var parts [3]int
	segments := strings.Split(version, ".")
	if len(segments) < 1 || len(segments) > 3 {
		return parts, false
	}
	for i := 0; i < len(segments); i++ {
		n, err := strconv.Atoi(segments[i])
		if err != nil || n < 0 {
			return parts, false
		}
		parts[i] = n
	}
	return parts, true
}

File: src/update-check.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { agentModeEnabled } from "./agent-mode.js";
import { Version, displayVersion, normalizeVersion, userAgent } from "./version.js";

const updateCheckRepo = "OpenVanta/cli";
const updateCheckIntervalMs = 24 * 60 * 60 * 1000;
const updateCheckHTTPTimeoutMs = 1500;
const updateCheckEnvDisable = "VANTA_NO_UPDATE";
const installScriptURL =
  "https://raw.githubusercontent.com/OpenVanta/cli/main/scripts/install.sh";

type UpdateCheckCache = {
  checked_at: string;
  latest_version: string;
};

type UpdateCheckOptions = {
  agentMode?: boolean;
  /** Injected for tests. */
  now?: () => Date;
  fetchImpl?: typeof fetch;
  stderrIsTTY?: boolean;
  env?: NodeJS.ProcessEnv;
};

let checkPromise: Promise<string | null> | null = null;

export function envTruthy(value: string | undefined): boolean {
  switch ((value ?? "").trim().toLowerCase()) {
    case "1":
    case "true":
    case "yes":
    case "on":
      return true;
    default:
      return false;
  }
}

export function isNewerVersion(latest: string, current: string): boolean {
  const latestParts = parseSemver(normalizeVersion(latest));
  const currentParts = parseSemver(normalizeVersion(current));
  if (!latestParts || !currentParts) {
    const a = normalizeVersion(latest);
    const b = normalizeVersion(current);
    return a !== b && a > b;
  }
  for (let i = 0; i < 3; i++) {
    if (latestParts[i]! > currentParts[i]!) return true;
    if (latestParts[i]! < currentParts[i]!) return false;
  }
  return false;
}

function parseSemver(version: string): [number, number, number] | null {
  const segments = version.split(".");
  if (segments.length < 1 || segments.length > 3) return null;
  const parts: [number, number, number] = [0, 0, 0];
  for (let i = 0; i < segments.length; i++) {
    if (!/^\d+$/.test(segments[i]!)) {
      return null;
    }
    const n = Number.parseInt(segments[i]!, 10);
    if (!Number.isFinite(n) || n < 0) {
      return null;
    }
    parts[i] = n;
  }
  return parts;
}

export function shouldNotifyUpdates(options: UpdateCheckOptions = {}): boolean {
  const env = options.env ?? process.env;
  if (!normalizeVersion(Version) || Version === "dev") {
    return false;
  }
  if (envTruthy(env[updateCheckEnvDisable])) {
    return false;
  }
  if (envTruthy(env.CI)) {
    return false;
  }
  if (agentModeEnabled(options.agentMode)) {
    return false;
  }
  const tty =
    options.stderrIsTTY !== undefined
      ? options.stderrIsTTY
      : Boolean(process.stderr.isTTY);
  if (!tty) {
    return false;
  }
  return true;
}

function updateCheckCachePath(): string {
  return join(homedir(), ".vanta", "update-check.json");
}

async function readUpdateCheckCache(): Promise<UpdateCheckCache | null> {
  try {
    const raw = await readFile(updateCheckCachePath(), "utf8");
    return JSON.parse(raw) as UpdateCheckCache;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      return null;
    }
    return null;
  }
}

async function writeUpdateCheckCache(cache: UpdateCheckCache): Promise<void> {
  const path = updateCheckCachePath();
  await mkdir(dirname(path), { recursive: true, mode: 0o700 });
  await writeFile(path, `${JSON.stringify(cache, null, 2)}\n`, { mode: 0o600 });
}

async function fetchLatestReleaseVersion(
  fetchImpl: typeof fetch,
): Promise<string> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), updateCheckHTTPTimeoutMs);
  try {
    const resp = await fetchImpl(
      `https://api.github.com/repos/${updateCheckRepo}/releases/latest`,
      {
        headers: {
          Accept: "application/vnd.github+json",
          "User-Agent": userAgent(),
        },
        signal: controller.signal,
      },
    );
    if (!resp.ok) {
      throw new Error(`github releases latest: ${resp.status}`);
    }
    const release = (await resp.json()) as { tag_name?: string };
    const tag = release.tag_name?.trim() ?? "";
    if (!tag) {
      throw new Error("github releases latest: empty tag_name");
    }
    return normalizeVersion(tag);
  } finally {
    clearTimeout(timer);
  }
}

export async function resolveLatestVersion(
  options: UpdateCheckOptions = {},
): Promise<string> {
  const now = options.now ?? (() => new Date());
  const fetchImpl = options.fetchImpl ?? fetch;

  const cached = await readUpdateCheckCache();
  if (cached?.latest_version && cached.checked_at) {
    const checkedAt = new Date(cached.checked_at).getTime();
    if (
      Number.isFinite(checkedAt) &&
      now().getTime() - checkedAt < updateCheckIntervalMs
    ) {
      return cached.latest_version;
    }
  }

  const latest = await fetchLatestReleaseVersion(fetchImpl);
  await writeUpdateCheckCache({
    checked_at: now().toISOString(),
    latest_version: latest,
  }).catch(() => {
    /* ignore cache write failures */
  });
  return latest;
}

/** Start a background latest-version check (no-op if notices are disabled). */
export function startBackgroundUpdateCheck(
  options: UpdateCheckOptions = {},
): void {
  if (!shouldNotifyUpdates(options)) {
    return;
  }
  if (checkPromise) {
    return;
  }
  checkPromise = resolveLatestVersion(options)
    .then((latest) => latest)
    .catch(() => null);
}

/** Await the background check (briefly) and print a notice if outdated. */
export async function finishBackgroundUpdateCheck(
  options: UpdateCheckOptions = {},
  write: (s: string) => void = (s) => process.stderr.write(s),
): Promise<void> {
  if (!checkPromise) {
    return;
  }

  const latest = await Promise.race([
    checkPromise,
    new Promise<null>((resolve) =>
      setTimeout(() => resolve(null), updateCheckHTTPTimeoutMs),
    ),
  ]);
  checkPromise = null;

  if (!latest || !isNewerVersion(latest, Version)) {
    return;
  }

  write(
    `\nA new version of vanta is available: ${displayVersion(latest)} (you have ${displayVersion(Version)})\nUpdate with:\n  curl -fsSL ${installScriptURL} | bash\n\n`,
  );
}

/** Test helper to clear in-flight check state. */
export function resetUpdateCheckStateForTests(): void {
  checkPromise = null;
}
